elasticsearch/manifests/kustomization.yaml

```diff
@@ -0,0 +1,22 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+
+resources:
+  - elasticsearch.yaml
+
+patches:
+  - target:
+      kind: StatefulSet
+      name: elasticsearch-master
+    patch: |-
+      # replace ELASTICSEARCH_CLUSTER_HOSTS 
+      - op: replace
+        path: /spec/template/spec/containers/0/env/7/value
+        value: "elasticsearch-master-hl,"
+      # replace ELASTICSEARCH_ADVERTISED_HOSTNAME
+      - op: replace
+        path: /spec/template/spec/containers/0/env/11/value
+        value: "$(MY_POD_NAME).elasticsearch-master-hl"
```

Comment:
I'm concerned these fixed positions (`7`, `11`) could present a very subtle/possibly undetected change if we bump the bitnami version and they e.g. insert a new env var just before `ELASTICSEARCH_CLUSTER_HOSTS` (so `ELASTICSEARCH_CLUSTER_HOSTS` is now at position `8`)

I thought to try an inline strategic merge, so we can patch by name:

```yaml
patches:
  - target:
      kind: StatefulSet
    patch: |-
      apiVersion: apps/v1
      kind: StatefulSet
      metadata:
        name: elasticsearch-master
      spec:
        template:
          spec:
            containers:
              - name: elasticsearch
                env:
                  - name: ELASTICSEARCH_CLUSTER_HOSTS 
                    value:  "elasticsearch-master-hl,"
                  - name: ELASTICSEARCH_ADVERTISED_HOSTNAME
                    value: "$(MY_POD_NAME).elasticsearch-master-hl"
```

But this spits out an error:

```
$ kustomize build 
Error: updating name reference in 'spec/template/spec/volumes/configMap/name' field of 'StatefulSet.v1.apps/elasticsearch-master.[noNs]': considering field 'spec/template/spec/volumes/configMap/name' of object StatefulSet.v1.apps/elasticsearch-master.[noNs]: expected sequence or mapping node
```

That I could only fix by patching the source:

```patch
diff --git a/elasticsearch/manifests/elasticsearch.yaml b/elasticsearch/manifests/elasticsearch.yaml
index 9c396d8..29e479e 100644
--- a/elasticsearch/manifests/elasticsearch.yaml
+++ b/elasticsearch/manifests/elasticsearch.yaml
@@ -472,7 +472,7 @@ spec:
           volumeMounts:
             - name: data
               mountPath: /bitnami/elasticsearch/data
-      volumes:
+      volumes: []
   volumeClaimTemplates:
     - metadata:
         name: "data"
```

:thinking: not sure if this is a bug somewhere or it's "working as intended" :tm:

Reply:
[This discussion](https://www.reddit.com/r/kubernetes/comments/113g3sz/help_with_kustomize_cleanest_way_to_replace_an/) 
proposes following solution:
```
 - patch: |-
  - op: test
    path: /spec/template/spec/containers/0/env/1/name
    value: ENV_VAR_TO_REPLACE
  - op: replace
    path: "/spec/template/spec/containers/0/env/1"
    value:
      name: ENV_VAR_TO_REPLACE
      value: "new-val"
```
where we basically ensure it is the correct env.

I will use it- otherwise, if those variables will have wrong value, ES will fail with rather cryptic error `failed to resolve publish address`.

Reply:
[Done](https://github.com/utilitywarehouse/shared-kustomize-bases/pull/16/commits/89ce0e844c27cf23325276e826062d336723d2ba)

Reply:
Hmm, on this, I would still go with the patch that Matt suggested. Even place it in a separate yaml.

Reply:
Added comment [here](https://github.com/utilitywarehouse/shared-kustomize-bases/pull/16/commits/5ef645f9a43d86f54f5812d919014d542f7976ae). For some reason, inline strategic merge doesn't work correctly for me as well

Reply:
> inline strategic merge doesn't work correctly for me as well

Which error were you running into? I found replace the `volumes` in the patch allows things to build correctly:

```yaml
patches:
  - target:
      kind: StatefulSet
    patch: |-
      apiVersion: apps/v1
      kind: StatefulSet
      metadata:
        name: elasticsearch-master
      spec:
        template:
          spec:
            containers:
              - name: elasticsearch
                env:
                  - name: ELASTICSEARCH_CLUSTER_HOSTS 
                    value:  "elasticsearch-master-hl,"
                  - name: ELASTICSEARCH_ADVERTISED_HOSTNAME
                    value: "$(MY_POD_NAME).elasticsearch-master-hl"
            # the manifests generated from the Helm chart have 'volumes: ` (i.e. nil).
            # Kustomize doesn't seem to like that, so make it explicitly empty instead
            volumes: []
```

Reply:
I'd say put it in a yaml file anyway, as I think it's more readable & explanatory that way

Reply:
@matthewhughes-uw 
This patch changes the order of the env variables- and it tries to use another env as value, but this env that is not set yet. 
![image](https://github.com/utilitywarehouse/shared-kustomize-bases/assets/32140673/85e99478-c0f6-4d36-8307-c05dd575fff5)
I suppose we could also add MY_POD_ENV in the patch in order to force the order, but I'm not sure whether we will be more readable at this point

Reply:
> @matthewhughes-uw This patch changes the order of the env variables- and it tries to use another env as value, but this env that is not set yet. 

:slightly_frowning_face: this looks to be working as intended https://github.com/kubernetes-sigs/kustomize/issues/3157#issuecomment-719767040

> I suppose we could also add MY_POD_ENV in the patch in order to force the order, but I'm not sure whether we will be more readable at this point

:thinking: yeah, I'm leaning towards the original patching approach being better at this point, otherwise we need a patch that sets the env vars _and_ sets an extra env var (`MY_POD_NAME`) _and_ adds the `volumes` change which just feels like quite a bit of noise. What do you think @sbuliarca ?